### DISCUSSION TITLE
Clearing the app's data

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13153] [Q&A]
Each time if I build and deploy my app the snippet of `controllers/tab1.js` creates a error msg `table tab2 has no column named testString in "REPLACE INTO tab2...`. To solve this I can clear the app's data and restart the app or by adding the code snippet of `alloy.js`. 

1) Now I add the snippet from `controllers/tab1.js` to some other custom controller, e.g. `controllers/mytest.js`. 
2) Empty the `controllers/tab1.js` file.
3) In this configuration I run into same issue described above when `saveTestString()` is executed, with the difference that the solution does not work at all.
4) Only if I add a record in `controllers/tab1.js` before I trying to add another records in `controllers/mytest.js`, then no error msg appears.

alloy.js
```
 var db = Ti.Database.open ('_alloy_');
 db.execute ('DROP TABLE IF EXISTS tab2;');
 db.close ();

```

controllers/tab1.js
```
(function() {
  let tab1 = Alloy.createModel('tab2', {"testString": "myString");
  tab1.save();
})();

function saveString(event){
 	Alloy.createController('myTest').getView().open();
}

```

views/index.xml
```
<Alloy>
	<TabGroup title="MyApp">
           <Require src="tab1" />
           <Require src="tab2" />
	</TabGroup>
</Alloy>
```
views/tab1.xml
```
<Alloy>
   <Tab id="tab1" title="Tab1">
    <Window id="window1" title="Window1" onClose="cleanup">
        <Menu>
        <MenuItem title="Save" onClick="saveString"/>
      </Menu>
    </Window>
   </Tab>
  </Alloy>
```
views/tab2.xml
```
<Alloy>
  <Collection src="tab2" />
   <Tab id="tab2" title="Tab2">
    <Window id="window2" title="Window2" onClose="cleanup">
    </Window>
   </Tab>
  </Alloy>
```

models/tab2.js
```
exports.definition = {
    config: {
        "columns": {
            "testString": "TEXT"
            "id": "INTEGER PRIMARY KEY AUTOINCREMENT"
        },
        "adapter": {
            "type": "sql",
            "collection_name": "tab2",
            "idAttribute": "id"
        }
     }
}
```
controllers/mytest.js
```
function saveTestString(){
  let tab1 = Alloy.createModel('tab2', {"testString": "myString");
  tab1.save();
}
```
views/mytest.xml
```
<Alloy>
    <Window id="saveTestString" title="New Test String" onClose="cleanup">
       <ActionBar id="actionbar" displayHomeAsUp="true" onHomeIconItemSelected="closeWindow" />
         <Menu>
           <MenuItem showAsAction="Ti.Android.SHOW_AS_ACTION_ALWAYS" title="Save" onClick="saveOpponent">
           </MenuItem>
        </Menu>
     </Window>
</Alloy>
```

**Comment by m1ga** [comment]
When I try to run your example code it just stops with
```
[ERROR]  Invalid <TabGroup> child type: Alloy.Require
[ERROR]  All <TabGroup> children must be <Tab> or <Menu> or <ActionBar>
[ERROR]  Alloy compiler failed
```
Which version of Alloy are you using?
And isn't the "error" just a warning? I remember those messages if you have a wrong migration file but it still worked. But it might be different in your case.

The "REPLACE" part is here: https://github.com/appcelerator/alloy/blob/28a239aa7fb1fdef2ac67e50d35780f5c9dd766d/Alloy/lib/alloy/sync/sql.js#L189 so it looks like it runs an `update/sync` and there it can't find the field. But I don't know why since I can't run the example.

**Reply by beppo-ivel** [reply]
> When I try to run your example code it just stops with

It is not a complete working example. Please see the edits in the start post. 

> And isn't the "error" just a warning?

The Error is the same as described [here](https://stackoverflow.com/questions/20565441/titanium-alloy-models-table-has-no-column)

> if you have a wrong migration file

I don't have migration file at all. But yes, u can say I don't have to clear app's data if I would define a migration file.

**Reply by m1ga** [reply]
so its just a dev thing on your device and if it is the same as that SO post the database was created before without the field. If so: just add your filed there or use something like https://stackoverflow.com/a/43524278 to add it later on.
All corrects answers are in you SO post :) It is in the device data once you've created it and the best way to update it is a migration or cleaning the data. If the drop doesn't work: are you creating the old structure again? Are you accessing it before the drop? 
I wouldn't add too much time into that problem: if it is a new app, just clean the data or uninstall the app and create a new one. If it is a live app you want to update: create a migration!

**Reply by beppo-ivel** [reply]
> I wouldn't add too much time into that problem: if it is a new app, just clean the data or uninstall the app and create a new one.

I agree.  But there is one thing left I don't understand. If I delete the code snippet in controllers/tab1.js I run always into the described error above when `saveTestString()` from `controllers/mytest.js` is executed. The code in the self-executed-function and `saveTestString` is the same but only controllers/tab1.js is able to execute it and `saveTestString()` only executes without any error if the self-executed-function creates a record.
